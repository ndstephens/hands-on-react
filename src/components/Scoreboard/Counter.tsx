import styled from '@emotion/styled';

interface Props {
  /**
   * Number in the counter
   */
  children: string;
}

export const Counter = ({ children }: Props) => {
  return <Frame>{children}</Frame>;
};

const Frame = styled.div`
  display: inline-block;
  padding: 0 0.3vw;
  color: #ec433c;
  border: 0.15vw inset;
  /* line-height: 2vw; */
  line-height: 1.6;
  letter-spacing: 0.08em;
  background-color: #333;
  text-shadow: 0 0 0.1vw currentColor;
`;
