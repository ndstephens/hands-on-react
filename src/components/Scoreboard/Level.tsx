import styled from '@emotion/styled';

interface Props {
  /**
   * Array of possible game levels
   */
  children: string[];
}

export const Level = ({ children }: Props) => {
  return (
    <Select>
      {children.map((item: string) => (
        <Option key={item} value={item}>
          {item}
        </Option>
      ))}
    </Select>
  );
};

const Select = styled.select`
  margin: 0;
  height: 2.5vw;
  border-radius: 0;
  border: 0.15vw solid;
  border-color: white #9e9e9e #9e9e9e white;
  background-color: #d1d1d1;
`;

const Option = styled.option`
  font-weight: normal;
  display: block;
  white-space: nowrap;
  min-height: 1.2em;
  padding: 0 0.2vw 0.2vw;
`;
