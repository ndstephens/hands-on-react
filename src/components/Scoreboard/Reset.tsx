import styled from '@emotion/styled';

import { useMouseDown } from '@/hooks/useMouseDown';

const BUTTON_ICON = 'ðŸ˜¬';
const BUTTON_ICON_MOUSE_DOWN = 'ðŸ˜²';

interface Props {
  /**
   * Reset action handler
   */
  onReset: () => void;
}

export const Reset = ({ onReset }: Props) => {
  const [mouseDown, onMouseDown, onMouseUp] = useMouseDown();

  return (
    <Button
      onMouseDown={onMouseDown}
      onMouseUp={onMouseUp}
      onMouseLeave={onMouseUp}
      onClick={onReset}
      data-testid="reset-button"
    >
      {mouseDown ? BUTTON_ICON_MOUSE_DOWN : BUTTON_ICON}
    </Button>
  );
};

const Button = styled.button`
  cursor: pointer;
  font-size: 1.5vw;
  font-weight: 700;
  border-width: 0.15vw;
  border-style: solid;
  border-color: white #9e9e9e #9e9e9e white;
  background-color: #d1d1d1;

  &:active {
    border-color: #9e9e9e white white #9e9e9e;
  }
`;
