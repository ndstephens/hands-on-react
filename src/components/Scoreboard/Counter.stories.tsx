import { ComponentMeta, ComponentStory } from '@storybook/react';

import { Counter } from './Counter';

export default {
  title: 'Scoreboard/Counter',
  component: Counter,
} as ComponentMeta<typeof Counter>;

const Template: ComponentStory<typeof Counter> = (args) => (
  <Counter {...args} />
);

export const CounterExample = Template.bind({});
CounterExample.args = {
  children: '10',
};
