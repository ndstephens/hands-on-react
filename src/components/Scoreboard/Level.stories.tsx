import { ComponentMeta, ComponentStory } from '@storybook/react';

import { Level } from './Level';

export default {
  title: 'Level',
  component: Level,
} as ComponentMeta<typeof Level>;

const Template: ComponentStory<typeof Level> = (args) => <Level {...args} />;

export const LevelExample = Template.bind({});
LevelExample.args = {
  children: ['beginner', 'intermediate', 'expert'],
};
