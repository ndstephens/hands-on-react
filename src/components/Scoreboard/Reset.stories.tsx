import { ComponentMeta, ComponentStory } from '@storybook/react';

import { Reset } from './Reset';

export default {
  title: 'Reset',
  component: Reset,
} as ComponentMeta<typeof Reset>;

const Template: ComponentStory<typeof Reset> = (args) => <Reset {...args} />;

export const ResetExample = Template.bind({});
