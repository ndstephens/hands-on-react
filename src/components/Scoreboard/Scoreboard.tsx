import styled from '@emotion/styled';

import { Counter } from './Counter';
import { Level } from './Level';
import { Reset } from './Reset';

export interface ScoreboardProps {
  /**
   * Timer
   */
  time: string;
  /**
   * Possible game scenarios
   */
  levels: string[];
  /**
   * Number of bombs in the field
   */
  mines: string;
  /**
   * Action handler when the GameReset button is clicked
   */
  onReset: () => void;
}

export const Scoreboard = ({
  time,
  levels,
  mines,
  onReset,
}: ScoreboardProps) => {
  return (
    <Wrapper>
      <Counter>{time}</Counter>
      <Level>{levels}</Level>
      <Reset onReset={onReset} />
      <Counter>{mines}</Counter>
    </Wrapper>
  );
};

const Wrapper = styled.div`
  display: flex;
  justify-content: space-between;
  padding-bottom: 2vw;
  width: max-content;
`;
