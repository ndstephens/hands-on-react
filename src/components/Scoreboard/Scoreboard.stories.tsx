import { ComponentMeta, ComponentStory } from '@storybook/react';

import { Scoreboard } from './Scoreboard';

export default {
  title: 'Scoreboard/Scoreboard',
  component: Scoreboard,
} as ComponentMeta<typeof Scoreboard>;

const Template: ComponentStory<typeof Scoreboard> = (args) => (
  <Scoreboard {...args} />
);

export const ScoreboardExample = Template.bind({});
ScoreboardExample.args = {
  time: '000',
  levels: ['beginner', 'intermediate', 'expert'],
  mines: '010',
};
