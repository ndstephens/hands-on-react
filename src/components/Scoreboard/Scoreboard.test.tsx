import { render } from '@testing-library/react';

// import { render, screen } from '@testing-library/react';
// import userEvent from '@testing-library/user-event';
import { Scoreboard, ScoreboardProps } from './Scoreboard';

describe('Scoreboard test cases', () => {
  // Default props
  const props: ScoreboardProps = {
    time: '000',
    levels: ['beginner', 'intermediate', 'expert'],
    mines: '010',
    onReset: () => null,
    // onChangeLevel={() => null}
  };

  it('Scoreboard renders correctly', () => {
    const { asFragment } = render(<Scoreboard {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });

  // it('Scoreboard select level handler check', () => {
  //   const onChange = jest.fn();

  //   render(
  //     <Scoreboard
  //       time="000"
  //       levels={['beginner', 'intermediate', 'expert']}
  //       bombs="010"
  //       onReset={() => null}
  //       onChangeLevel={onChange}
  //     />
  //   );

  //   userEvent.selectOptions(screen.getByRole('combobox'), 'expert');

  //   expect(screen.getByRole('option', { name: 'expert' })).toBeEnabled();

  //   expect(onChange).toHaveBeenCalled();
  // });
});
