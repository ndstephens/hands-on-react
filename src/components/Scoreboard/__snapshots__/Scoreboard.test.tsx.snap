// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scoreboard test cases Scoreboard renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1xe3g8o"
  >
    <div
      class="css-1tfs0js"
    >
      000
    </div>
    <select
      class="css-55qvia"
    >
      <option
        class="css-v3kp7e"
        value="beginner"
      >
        beginner
      </option>
      <option
        class="css-v3kp7e"
        value="intermediate"
      >
        intermediate
      </option>
      <option
        class="css-v3kp7e"
        value="expert"
      >
        expert
      </option>
    </select>
    <button
      class="css-1cjjgsg"
      data-testid="reset-button"
    >
      ðŸ˜¬
    </button>
    <div
      class="css-1tfs0js"
    >
      010
    </div>
  </div>
</DocumentFragment>
`;
